lib/zip/crypto/aes_encryption.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+require 'openssl'
+
+module Zip
+  module AESEncryption # :nodoc:
+    VERIFIER_LENGTH = 2
+    BLOCK_SIZE = 16
+    AUTHENTICATION_CODE_LENGTH = 10
+
+    def initialize(password, strength)
+      @password = password
+      @strength = strength
+
+      n = @strength + 1
+      @headers = {
+        bits: 64 * n,
+        key_length: 8 * n,
+        mac_length: 8 * n,
+        salt_length: 4 * n
+      }
+      @counter = 0
+    end
+
+    def header_bytesize
+      @headers[:salt_length] + VERIFIER_LENGTH
+    end
+
+    def gp_flags
+      0
+    end
+  end
+
+  class AESDecrypter < Decrypter # :nodoc:
+    include AESEncryption
+
+    def decrypt(encrypted_data)
+      amount_to_read = encrypted_data.size
+      decrypted_data = +''
+
+      while amount_to_read > 0
+        @cipher.iv = [@counter + 1].pack('Vx12')
+        begin_index = BLOCK_SIZE * @counter
+        end_index = BLOCK_SIZE * @counter + [BLOCK_SIZE, amount_to_read].min - 1
+        decrypted_data << @cipher.update(encrypted_data[begin_index..end_index])
+        amount_to_read -= BLOCK_SIZE
+        @counter += 1
+      end
+
```

Comment:
We are missing the check of the CRC, we should check that the decrypted content hasn't been tampered.

Reply:
I've tried to integrate the AES integrity check, but it doesn't work.
If you have an idea, I'd love to hear it.

https://github.com/jplot/rubyzip/commit/21bdb688f2fc11a98d7855b78ace04509a2e49ea